### DISCUSSION TITLE
Converting from .jpg to .pdf yields blank pages

### DISCUSSION THREAD

**Discussion by grrigore** [ImageMagick/ImageMagick#4580] [Help] [answered]
I have some problems converting 180 JPG files into a single PDF.

First, I resize the JPG files using `mogrify -resize 75% *.jpg` and the result looks fine.
After that, I run `convert *.jpg converted.pdf` and it works fine for the first 3-4 (out of 180) JPG files.

I don't understand why some of the images are properly converted, while most of them are blank.

Thank you and let me know if you need some extra information.

**Comment by fmw42** [comment]
The issue is that you load all your image into memory when using convert and running out of RAM. Whereas, mogrify loads the images one at a time processes it and saves it then goes to the next image after deleting the first.

**Reply by grrigore** [reply]
I'm on a 16GB MacBook. I don't think that's the problem. One image is around 1MB in size.

**Reply by grrigore** [reply]
I've tried with only one image and got a blank page.

**Reply by fmw42** [reply]
You may be running out of RAM or running out of /tmp space on disk.

I was just explaining how convert and mogrify differ in use of RAM.

**Reply by fmw42** [reply]
Check your /tmp directory and be sure it is not full. Remove all ImageMagick tmp files in that directory.

**Reply by grrigore** [reply]
I don't think this is related to any of this. Maybe the jpg files are different somehow. Does it matter how the images are named? The first three images seem to work just fine (named: XXXX V1-000.jpg, XXXX V1-001.jpg, XXXX V1-002.jpg) but for XXXX V1-181.jpg it won't work.

**Reply by fmw42** [reply]
Names you have should be fine.  Did you check your /tmp directory?  Did it have any ImageMagick tmp files?

**Reply by grrigore** [reply]
No, I don't think so. Nothing that contains ImageMagick or images names or something related.

**Reply by fmw42** [reply]
How many images are you loading and what are the image dimensions and how much RAM do you have? You may be running out of RAM

Why don't you just do

`mogrify -format pdf -resize 75% *.jpg`

That will save them as pdf files directly from mogrify.  You can also add -path path_to/existing_directory to save them to some already existing (empty) directory if you do not want to mix them with your input jpg files.

**Reply by grrigore** [reply]
There are 180 images with sizes from 800KB to 1.2MB. I have 16GB of RAM. You can assume 200 images of 2MB, I think 16GB are still enough, right?

I've tried `mogrify -format pdf -resize 75% *.jpg` but still getting single blank page PDFs for most of them. Is there a way to obtain some logs?

**Reply by fmw42** [reply]
If mogrify fails, then see if you can manually convert some of the bad ones with convert using one image at a time. Try converting pdf then try again converting to say png or gif. Do they convert? Just checking if it is the input jpg files.

Also I wanted the dimensions, not the file sizes, since jpg is compressed and ImageMagick will decompress. So I have to compute RAM usage from the actual dimensions.

**Reply by grrigore** [reply]
So I tried converting .jpg -> .pdf and .pdf -> .png using convert. The .pdf one si blank, while the .png one is the image but displayed 3 times in a row.

I've used these commands:

```
convert test.jpg test.pdf
convert test.pdf test.png
```

Images are 2037 × 3056 (original size).

**Reply by grrigore** [reply]
Might have misunderstood that. I've converted the jpg to pdf using macOS quick actions and then using convert I converted it to png. Another thing is that if I use the png format the pdf is generated ok.

**Reply by fmw42** [reply]
You might have jpg files compressed with older compression that libjpeg does not support. Though if the JPG to PNG conversion is OK, then that is not the issue. Please use convert to convert to PDF, to PNG and to GIF. Use the same command line. Do not use some action from some other tool.

**Comment by snibgo** [comment]
> There are 180 images with sizes from 800KB to 1.2MB. I have 16GB of RAM. You can assume 200 images of 2MB, I think 16GB are still enough, right?

Wrong. Filesizes say nothing about how much memory is needed. Memory depends on the number of pixels. How may pixels, in total, are in your JPEGs?

**Reply by grrigore** [reply]
They are 2037 × 3056.

**Reply by fmw42** [reply]
I believe that is almost 45 GB and you only have 16 GB. You are running out of RAM as I said before.

180\*2037\*3056\*4bytes/pixel = 44.8 GB 

4bytes = 4*8 bits = 24-bit color image

**Reply by grrigore** [reply]
What about it not working on a single image?

**Reply by fmw42** [reply]
Did you do the conversion tests that I requested?  What happened?

**Reply by grrigore** [reply]
Yes, check [here](https://github.com/ImageMagick/ImageMagick/discussions/4580#discussioncomment-1791630)

**Reply by fmw42** [reply]
Sorry, I see those tests above.  Can you post an example image?

What do you get from

`convert -version`

**Reply by grrigore** [reply]
`convert -version` yields:

```
Version: ImageMagick 7.1.0-17 Q16-HDRI x86_64 2021-12-04 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(5.0)
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib
Compiler: gcc (4.2)
```

**Reply by fmw42** [reply]
You are on ImageMagick 7. You should be using magick not convert.

If I do

`magick 145692032-1e3ee796-6a51-4b35-9ea0-bdb5c1e41ae5.jpg 145692032-1e3ee796-6a51-4b35-9ea0-bdb5c1e41ae5.pdf`

The PDF is completely white using IM 7.1.0-17 on Mac OSX Sierra

However, if I use IM 6.9.12-32, it works fine

`convert 145692032-1e3ee796-6a51-4b35-9ea0-bdb5c1e41ae5.jpg 145692032-1e3ee796-6a51-4b35-9ea0-bdb5c1e41ae5.pdf`

**So it appears that you have found a bug in IM 7 or something odd about your JPG files.**

**So one of the developers will have to look into this.**

**Reply by grrigore** [reply]
I see. Thank you! I will try to use version 6.9.12-32. Thank you so much for your time and support!

**Comment by Jmuccigr** [comment]
Maybe this isn't the right tool. Try img2pdf which allows you to choose the page and margin sizes. I use it all the time with a large number of image and have no problems.

**Reply by grrigore** [reply]
Yes, img2pdf worked, but the PDF size was quite big so I thought I'd give this one more try.

**Reply by Jmuccigr** [reply]
Yes, it's big because your files are big. :-) There's very little file overhead with this method, so if you want a smaller PDF then you need smaller images. You could use IM to convert the images to the size you want and then feed those to img2pdf, though you seem to be converting them already.

**Reply by grrigore** [reply]
Yes, I already reduce the size of the files. I don't know exactly why, but if I use an online tool (ilovepdf) I get files around 50MB instead of approximately 120MB ones.. maybe there are some settings I don't set.

**Reply by Jmuccigr** [reply]
@grrigore, my point is that the PDF is only as big as the images when made with img2pdf. So if you're getting smaller files with some other method, then that method involves more compressed or just smaller images.

**Reply by grrigore** [reply]
Yes, I guess the other tools do compress the pdf.

**Reply by Jmuccigr** [reply]
> Yes, I guess the other tools do compress the pdf.

They are almost surely compressing **the images** further. You really can't get much better compression on a jpeg without decreasing the quality of the image itself.

**Comment by urban-warrior** [comment, marked as answer]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/download/beta/ by sometime tomorrow.

**Reply by grrigore** [reply]
Thank you! I've tried it and it works.
